python/lsst/ts/atspectrograph/atspec_csc.py

```diff
@@ -102,26 +102,64 @@ def __init__(self, config_dir=None, initial_state=salobj.State.STANDBY,
+            # range.
+            state = await self.model.query_gs_status(self.want_connection)
+            self.log.debug(f"query_gs_status: {state}")
+            if state[1] < 0.:
+                self.log.warning("Linear stage out of range. Homing.")
+                await self.home_element(query="query_gs_status",
+                                        home="init_gs",
+                                        report="reportedLinearStagePosition",
+                                        inposition="linearStageInPosition",
+                                        report_state="lsState")
+            else:
+                self.evt_reportedLinearStagePosition.set_put(position=state[1])
+        except Exception as e:
+            self.evt_errorCode.set_put(errorCode=CONNECTION_ERROR,
+                                       errorReport=f"Cannot get information from model for "
+                                                   f"linear stage.",
+                                       traceback=traceback.format_exc())
+            self.fault()
+            raise e
+
+        try:
+
+            # Check/Report Filter Wheel position.
+            state = await self.model.query_fw_status(self.want_connection)
+            self.log.debug(f"query_fw_status: {state}")
+            filter_name = str(list(self.model.filters.keys())[int(state[1])])
+            self.evt_reportedFilterPosition.set_put(position=int(state[1])+1,
+                                                    name=filter_name)
+        except Exception as e:
+            self.evt_errorCode.set_put(errorCode=CONNECTION_ERROR,
+                                       errorReport=f"Cannot get information from model for "
+                                                   f"filter wheel.",
+                                       traceback=traceback.format_exc())
+            self.fault()
+            raise e
+
+        try:
+            # Check/Report Grating Wheel position.
+            state = await self.model.query_gw_status(self.want_connection)
+            self.log.debug(f"query_gw_status: {state}")
+            grating_name = str(list(self.model.gratings.keys())[int(state[1])])
+            self.evt_reportedDisperserPosition.set_put(position=int(state[1])+1,
+                                                       name=grating_name)
+        except Exception as e:
+            self.evt_errorCode.set_put(errorCode=CONNECTION_ERROR,
+                                       errorReport=f"Cannot get information from model for "
+                                                   f"grating wheel.",
+                                       traceback=traceback.format_exc())
+            self.fault()
+            raise e
 
         self._health_loop = asyncio.ensure_future(self.health_monitor_loop())
 
-        await super().end_enable(id_data)
+        await super().end_enable(data)
 
-    async def end_disable(self, id_data):
+    async def end_disable(self, data):
```

Comment:
Don't you also want to wait for the loop any time you leave the enabled state?

I *strongly* suggest putting this in `handle_summary_state` instead and calling it whenever the state leaves "enabled".

That said, it is more common to leave connections to low-level controllers and such running when enabled or disabled (which is why `BaseCsc` has an `enabled_or_disabled` property) and kill them when in any other state. Are you sure you want the health loop to exit when state is disabled?